internal/hcs/errors.go

```diff
@@ -99,48 +103,85 @@ type ErrorEvent struct {
 	EventID    uint16 `json:"EventId,omitempty"`
 	Flags      uint32 `json:"Flags,omitempty"`
 	Source     string `json:"Source,omitempty"`
-	//Data       []EventData `json:"Data,omitempty"`  // Omit this as HCS doesn't encode this well. It's more confusing to include. It is however logged in debug mode (see processHcsResult function)
-}
 
-type hcsResult struct {
-	Error        int32
-	ErrorMessage string
-	ErrorEvents  []ErrorEvent `json:"ErrorEvents,omitempty"`
+	// Omit this as HCS doesn't encode this well. It's more confusing to include.
+	// It is however logged in during [processHcsResult] errors.
+	//Data       []EventData `json:"Data,omitempty"`
 }
 
 func (ev *ErrorEvent) String() string {
-	evs := "[Event Detail: " + ev.Message
+	sb := strings.Builder{}
+	ev.writeTo(&sb)
+	return sb.String()
+}
+
+func (ev *ErrorEvent) writeTo(b *strings.Builder) {
+	// rough wag at needed length
+	b.Grow(64 + len(ev.Message) + len(ev.StackTrace) + len(ev.Provider) + len(ev.Source))
+
+	// [strings.Builder] Write* functions always return nil errors
+	_, _ = b.WriteString("[Event Detail: " + ev.Message)
 	if ev.StackTrace != "" {
-		evs += " Stack Trace: " + ev.StackTrace
+		_, _ = b.WriteString(" Stack Trace: " + ev.StackTrace)
 	}
 	if ev.Provider != "" {
-		evs += " Provider: " + ev.Provider
+		_, _ = b.WriteString(" Provider: " + ev.Provider)
 	}
 	if ev.EventID != 0 {
-		evs = fmt.Sprintf("%s EventID: %d", evs, ev.EventID)
+		fmt.Fprintf(b, " EventID: %d", ev.EventID)
```

Comment:
We should perhaps use the same semantics for writing i.e. `WriteString`.

Reply:
wed have to call `b.WriteString(fmt.Sprintf(" EventID: %d", ev.EventID))` to handle the formatting, which feels more convoluted than `Fprintf`